packages/components/src/modifiers/hds-code-editor.ts

```diff
@@ -0,0 +1,189 @@
+
+  onInput: HdsCodeEditorSignature['Args']['Named']['onInput'];
+
+  observer!: IntersectionObserver;
+
+  modify(
+    element: HTMLElement,
+    positional: PositionalArgs<HdsCodeEditorSignature>,
+    named: NamedArgs<HdsCodeEditorSignature>
+  ): void {
+    assert('HdsCodeEditor must have an element', element);
+    // the intersection observer makes loading unreliable in tests
+    if (config.environment === 'test') {
+      this.setupTask.perform(element, positional, named);
+    } else {
+      this.observer = new IntersectionObserver(
+        (entries) => {
+          entries.forEach((entry) => {
+            if (entry.isIntersecting && this.setupTask.performCount === 0) {
+              this.setupTask.perform(element, positional, named);
+            }
+          });
+        },
+        {
+          rootMargin: LOADER_HEIGHT,
+        }
+      );
+
+      this.observer.observe(element);
+    }
+  }
+
+  willRemove() {
+    this.observer.disconnect();
+  }
+
+  loadLanguageTask = task(
+    { drop: true },
+    async (language?: HdsCodeEditorLanguages) => {
+      if (language === undefined) {
+        return;
+      }
+
+      let module: CodemirrorLanguageModule | null = null;
+      let languageFunction = null;
+
+      switch (language) {
+        case 'go':
+          module = await import('@codemirror/lang-go');
+          languageFunction = (module as CodemirrorGoModule).go;
+          break;
+        case 'json':
+          module = await import('@codemirror/lang-json');
+          languageFunction = (module as CodemirrorJsonModule).json;
+          break;
+        case 'sql':
+          module = await import('@codemirror/lang-sql');
+          languageFunction = (module as CodemirrorSqlModule).sql;
+          break;
+        case 'hcl':
+          module = await import('codemirror-lang-hcl');
+          languageFunction = (module as CodemirrorHclModule).hcl;
+          break;
+        default:
+          throw new Error(`Language ${language} is not supported`);
+      }
+
+      return languageFunction();
+    }
+  );
+
+  setupTask = task(
+    { drop: true },
+    async (
+      element: HTMLElement,
+      _positional: PositionalArgs<HdsCodeEditorSignature>,
+      named: NamedArgs<HdsCodeEditorSignature>
+    ) => {
+      const { onInput, onSetup, language, value } = named;
+
+      const [
+        {
+          EditorView,
+          keymap,
+          lineNumbers,
+          highlightActiveLineGutter,
+          highlightSpecialChars,
+          highlightActiveLine,
+        },
+        { EditorState },
+        { defaultKeymap, history, historyKeymap },
+        { bracketMatching, syntaxHighlighting },
+      ] = await Promise.all([
+        import('@codemirror/view'),
+        import('@codemirror/state'),
+        import('@codemirror/commands'),
+        import('@codemirror/language'),
+      ]);
+
+      this.onInput = onInput;
+
+      const languageExtension = await this.loadLanguageTask.perform(language);
```

Comment:
suggestion: we have a chain of promises here: the perform action as well as the async loading of modules within the `loadLanguageTask`. I suggest we add some error handling to prevent a JS crash in case things go wrong along the way.